storage/oasis/client.go

```diff
@@ -4,107 +4,95 @@ package oasis
-// If `skipChainContextCheck` is true, it also checks that
-// the RPC endpoint in `network` serves the chain context specified
-// in `network`.
-func NewClientFactory(ctx context.Context, network *config.Network, skipChainContextCheck bool) (*ClientFactory, error) {
-	var conn connection.Connection
-	var err error
-	if skipChainContextCheck {
-		conn, err = connection.ConnectNoVerify(ctx, network)
-	} else {
-		conn, err = connection.Connect(ctx, network)
-	}
-	if err != nil {
-		return nil, err
+func NewConsensusClient(ctx context.Context, sourceConfig *config.SourceConfig) (*ConsensusClient, error) {
+	switch {
+	case sourceConfig.History != nil:
+		nodeApi, err := history.NewHistoryConsensusApiLite(
+			ctx,
+			history.DefaultChains[sourceConfig.History.ChainName],
+			sourceConfig.History.Nodes,
+			sourceConfig.FastStartup,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("instantiating history consensus API lite: %w", err)
+		}
+		network := sdkConfig.DefaultNetworks.All[sourceConfig.History.ChainName]
+		return &ConsensusClient{
+			nodeApi: nodeApi,
+			network: network,
+		}, nil
+	case sourceConfig.SingleNetwork != nil:
+		sdkConn, err := history.SDKConnect(
+			ctx,
+			sourceConfig.SingleNetwork.ChainContext,
+			sourceConfig.SingleNetwork.Node,
+			sourceConfig.FastStartup,
+		)
+		if err != nil {
+			return nil, err
+		}
+		coreClient := sdkConn.Consensus()
+		nodeApi := damask.NewDamaskConsensusApiLite(coreClient)
+		network := &sdkConfig.Network{
+			ChainContext: sourceConfig.SingleNetwork.ChainContext,
+			// no RPC
+			// no Denomination
+			// no ParaTimes
+		}
+		return &ConsensusClient{
+			nodeApi: nodeApi,
+			network: network,
+		}, nil
+	default:
+		return nil, fmt.Errorf("source not configured")
 	}
-
-	return &ClientFactory{
-		connection: &conn,
-		network:    network,
-	}, nil
 }
 
-// Consensus creates a new ConsensusClient.
-func (cf *ClientFactory) Consensus() (*ConsensusClient, error) {
-	// TODO: Remove the hardcoded values in next block once indexer config supports multiple nodes.
-	// The plan is to introduce a new implementation of ConsensusApiLite that
-	// will forward each call to either CobaltConsensusApiLite or DamaskConsensusApiLite,
-	// depending on the `height` parameter in the call.
-	var nodeApi nodeapi.ConsensusApiLite
-	if cf.network.ChainContext == "53852332637bacb61b91b6411ab4095168ba02a50be4c3f82448438826f23898" {
-		// Cobalt mainnet.
-		creds := credentials.NewTLS(&tls.Config{MinVersion: tls.VersionTLS12})
-		dialOpts := []grpc.DialOption{grpc.WithTransportCredentials(creds)}
-		grpcConn, err := cmnGrpc.Dial(cf.network.RPC, dialOpts...)
+func NewRuntimeClient(ctx context.Context, sourceConfig *config.SourceConfig, runtime analyzer.Runtime) (*RuntimeClient, error) {
+	var sdkClient connection.RuntimeClient
+	var sdkNet *sdkConfig.Network
+	switch {
+	case sourceConfig.History != nil:
+		// Use only the current node for runtime interaction for now.
```

Comment:
Can you open a clickup ticket for the remaining work please?
- encode lowest runtime round for each runtime and chain (Damask etc) in `history.DefaultChains`
- support per-runtime node configs (e.g. using a separate node for Sapphire)

I think a single ticket/PR can handle it, WDYT?

Reply:
k

Reply:
https://app.clickup.com/t/866a0fpr3